### PR TITLE
Schema for package config/perms workflows (SYN-4203)

### PR DESCRIPTION
**Pull request by Cisphyx**

### CODE REVIEW

File: synapse/lib/storm.py [hunk 1]
```diff
@@ -230,10 +230,25 @@
...
         },
+        'configvars': {
+            'type': 'array',
+            'items': {
+                'type': 'object',
+                'properties': {
+                    'name': {'type': 'string'},
+                    'varname': {'type': 'string'},
+                    'desc': {'type': 'string'},
+                    'default': {'type': 'string'},
```
Comment (invisig0th) [inline synapse/lib/storm.py:246]: This will probably need to potentially be almost any type?

File: synapse/lib/storm.py [hunk 2]
```diff
@@ -230,10 +230,25 @@
...
+                    'workflowconfig': {'type': 'boolean'},
```
Comment (invisig0th) [inline synapse/lib/storm.py:247]: We may need a way to permit different scopes in an extensible way. global / self ( and maybe a future "role"? )

File: synapse/tests/files/stormpkg/testpkg.yaml [hunk 3]
```diff
@@ -22,6 +22,24 @@ docs:
...
+    - perm: [ power-ups, testpkg, user ]
+      gate: cortex
+      desc: Controls user access to testpkg.
+      workflowconfig: true
+
+configvars:
+    - name: API key
+      varname: testpkg:apikey
+      desc: API key to use for querying the testpkg API.
+      workflowconfig: true
```
Comment (invisig0th) [inline synapse/tests/files/stormpkg/testpkg.yaml:35]: What does workflowconfig currently denote?

Comment (Cisphyx) [reply synapse/tests/files/stormpkg/testpkg.yaml:35]: Whether there should be a workflow for configuring the var. Could just skip it and say that anything in these lists gets a workflow.

File: synapse/lib/storm.py [hunk 4]
```diff
@@ -230,10 +230,32 @@
...
                 },
                 'required': ['perm', 'desc', 'gate'],
             },
         },
+        'configvars': {
+            'type': 'array',
+            'items': {
+                'type': 'object',
+                'properties': {
+                    'name': {'type': 'string'},
```
Comment (invisig0th) [inline synapse/lib/storm.py:250]: could we add some type knowledge ( either embedded JSON Schema or maybe cortex model types? ) to this so we can have auto normalization/enforcement?

File: synapse/tests/files/stormpkg/testpkg.yaml [hunk 5]
```diff
@@ -22,6 +22,25 @@ docs:
     - title: Foo Bar
       path: docs/foobar.md
 
+perms:
+    - perm: [ power-ups, testpkg, user ]
+      gate: cortex
+      desc: Controls user access to testpkg.
+      scopes: [ global, self ]
```
Comment (invisig0th) [inline synapse/tests/files/stormpkg/testpkg.yaml:29]: What does it mean for perms declarations to have scope?

**Review by invisig0th — DISMISSED** [review]

File: synapse/tests/files/stormpkg/testpkg.yaml (hunk 5 above)
Comment (Cisphyx) [reply synapse/tests/files/stormpkg/testpkg.yaml:29]: Scope should be on vars not perms

File: synapse/lib/storm.py [hunk 6]
```diff
@@ -318,6 +341,12 @@
             'additionalProperties': True,
             'required': ['form'],
         },
+        'configvartype': {
+            'anyOf': [
+                {'type': 'array', 'items': {'$ref': '#/definitions/configvartype'}},
+                {'enum': list(s_datamodel.Model().types)},
```
Comment (vEpiphyte) [inline synapse/lib/storm.py:347]: This gets us the following core types; but doesn't get the rest of our model types.

```
>>> import synapse.datamodel as s_datamodel
>>> list(s_datamodel.Model().types)
['int', 'float', 'range', 'str', 'hex', 'bool', 'time', 'duration', 'ival', 'guid', 'syn:tag:part', 'syn:tag', 'comp', 'loc', 'ndef', 'array', 'edge', 'timeedge', 'data', 'nodeprop', 'hugenum', 'taxon', 'taxonomy', 'velocity']
```

@invisig0th were you intending for things like `inet:url` or `inet:fqdn` to be valid types here?

File: synapse/lib/storm.py [hunk 7]
```diff
@@ -230,10 +230,33 @@
...
+                    'type': {'$ref': '#/definitions/configvartype'},
+                    'scopes': {
+                        'type': 'array',
+                        'items': {
+                            'type': 'string',
+                            'enum': ['global', 'self']
+                        },
+                    },
+                },
+                'required': ['name', 'varname', 'desc', 'type'],
```
Comment (vEpiphyte) [inline synapse/lib/storm.py:257]: are there instances where scope would not matter?

Comment (Cisphyx) [reply synapse/lib/storm.py:257]: probably not, i'll make it required

File: synapse/lib/storm.py (hunk 6 above)
Comment (Cisphyx) [reply synapse/lib/storm.py:347]: updated to validate types on add

**Review by vEpiphyte — APPROVED** [review]

**Pull request merged**